docs/oshu.1.in
.TH OSHU "1" "November 2017" "oshu! @VERSION@" "User Manual"

.SH NAME
oshu \- Lightweight port of the osu! rhythm game

.SH SYNOPSIS
.B oshu
\fIBEATMAP\fR.osu

.SH DESCRIPTION
.PP
Play a \fB.osu\fR beatmap file. Most linked files are automatically found from
the data inside the beatmap file.
.PP
On startup, oshu moves to the real directory of the beatmap, where it expects
to find the audio file. This lets you symlink your .osu files without bothering
with the related audio and picture files. In more technical terms, symbolic
terms are dereferenced.
.PP
\fB.osz\fR files are not supported, but since they're just fancy ZIP files, you
just have to unzip them into a sub-directory. Be careful though, it won't
create the sub-directory by itself if you naively unzip it.
.TP
\fB\-v, \-\-verbose\fR
Increase the verbosity. This will print debugging messages, which may contain
useful information for understanding errors when fixing issues. Mention it
twice and you'll get verbose debugging messages.
.TP
\fB\-h, \-\-help\fR
Show a brief help message.
.TP
\fB\-\-autoplay\fR
Perform a perfect run on the beatmap without any user interaction.
.TP
\fB\-\-pause\fR
Start the game in a paused state. Might be useful when you're starting the game
from a terminal as you won't be holding your mouse when the game starts.

.SH CONTROLS
.PP
The game is controlled from the keyboard and mouse. The keys are based on
osu!'s default settings, and rely on the \fIphysical\fR position on a QWERTY
keyboard. If you use a different keyboard layout, the keys are going to be at
the same position, even if the letters are different.
.SS Standard mode
.PP
Here are the official keys for the osu!standard mode:
.TP
\fBZ, X, mouse click\fR
Hit an object.
.SS Common keys
.PP
The following keys are common to all modes:
.TP
\fBEscape\fR
Toggle pause.
.TP
\fBPage up\fR
Rewind the song by 10 seconds, and leave a 1-second break.
.TP
\fBPage down\fR
Forward the song by 20 seconds.
.SS Pause
.PP
When the game is paused, your local keyboard layout is used.
These keys also apply to the score screen at the end.
.TP
\fBQ\fR
Quit the game.

.SH CONFIGURATION
.PP
oshu! currently doesn't read a configuration file, but reads the environment.
You may configure oshu! from your .bashrc, or by creating a wrapper script
\fIexport\fRing your variables.
.PP
Here's the list of supported environment variables:
.TP
\fBOSHU_WINDOW_SIZE\fR
This variable specifies the default window size when the game start. Its value
must conform to the format \fI960x720\fR, where 960 is the width in pixels, and
720 the height in pixels. When omitted or invalid, the game starts with a
fallback default size of 960x720.
.TP
\fBOSHU_QUALITY\fR
This variables controls the visual effects and other rendering quality
settings. It may take one of \fIlow\fR, \fImedium\fR, and \fIhigh\fR. The
default is \fIhigh\fR.

.SH AUTHOR
Written by Frédéric Mangano-Tarumi <fmang+oshu at mg0 fr>.

.SH BUGS
Report bugs at <https://github.com/fmang/oshu/issues>.
.br
Feel also free to join the official IRC channel, #oshu at chat.freenode.net.


CHANGELOG
Changelog
=========

1.6.0 - TBA
-----------

Featuring an improved user interface.

The game now displays the beatmap's metadata in-game, provides a clearer pause
screen and ends with a score screen. In the short term, the console should only
be targeted at the most curious users.

oshu! now supports a few performance settings configurable from the
*OSHU_QUALITY* and *OSHU_WINDOW_SIZE* environment variables. This may be
required for the game to be playable on low hardware, and in particular the
Raspberry Pi. More details in the man page.

1.5.0 - 2017-11-25
------------------

New fancy graphics!

This marks a milestone in the oshu! project, as the graphics engine was the
last remnant of the initial version. The new visuals uses accelerated 2D
graphics to render antialiased shapes and translucent objects. This looks nice
but is a significant load compared to the previous wireframe graphics. All the
objects are dynamically drawn using the cairo vector graphics library.

1.4.0 - 2017-11-11
------------------

Introducing seeking ability, to let you replay specific parts of a beatmap.

Overall, there are no revolutionary features, but the graphics and game modules
have been reorganized and improved for more stability, and with the future
modes in mind. Sliders and hit sounds have been tweaked to get closer to the
official osu! client. The command-line output is nicer, with colors instead of
crude log lines.

The keyboard controls have changed a bit, check out the man page.

1.3.0 - 2017-10-14
------------------

Hit sounds!

While the hi-hat was a perfectly decent generic soft sound, exploiting the full
audio sample range gives a completely different experience. Now you'll hear
rhythmic drums if you play well, and somethings other nice sound effects.

With this release comes a new parser, much more accurate than the previous one.
When a beatmap contains an error, the error is displayed and everyone is
invited to report them as oshu issues, because it's likely something is
missing.

The audio backend is also completely revamped to feature a flexible way to play
many sound effects at once. It should also be much more readable and organized.

1.2.0 - 2017-09-16
------------------

Background pictures!

The overall graphics engine now uses more precise positions, yielding smoother
curves, and also uses a finer game clock, while ensuring a constant frame rate.

It's mostly a technical release though, with a more generic game module
refactored to support more game modes in the future. More refactoring to come
in the next release.

1.1.0 - 2017-09-02
------------------

Full slider support!

The game is now quite playable, and starts getting fun.

Extra features:

- Resizable window.
- Slightly nicer graphics, even though it's still wireframe.
- Dereference the symlinked beatmap files.
- Automatic pause when the window focus is lost.
- `--pause` option to start the game paused.
- Print a crude score on the terminal output at the end of songs.

1.0.0 - 2017-08-07
------------------

This is first working prototype of the project. Let's call it a
proof-of-concept. It is basically playable even though it lacks major features
like sliders and scoring.

Here's what we got so far:

- Extremely basic gameplay.
- Overall support of most of the official beatmaps.


src/oshu.cc
/**
 * \file oshu.cc
 *
 * \brief
 * Main entry-point for the game.
 *
 * Interpret command-line arguments and spawn everything.
 */

#include "../config.h"

#include "log.h"
#include "game/game.h"

extern "C" {
#include <libavutil/log.h>
}

#include <SDL2/SDL.h>

#include <errno.h>
#include <getopt.h>
#include <signal.h>
#include <stdio.h>
#include <unistd.h>

enum option_values {
	OPT_AUTOPLAY = 0x10000,
	OPT_HELP = 'h',
	OPT_PAUSE = 0x10001,
	OPT_VERBOSE = 'v',
};

static struct option options[] = {
	{"autoplay", no_argument, 0, OPT_AUTOPLAY},
	{"help", no_argument, 0, OPT_HELP},
	{"pause", no_argument, 0, OPT_PAUSE},
	{"verbose", no_argument, 0, OPT_VERBOSE},
	{0, 0, 0, 0},
};

static const char *flags = "vh";

static const char *usage =
	"Usage: oshu [-v] BEATMAP.osu\n"
	"       oshu --help\n"
;

static const char *help =
	"Options:\n"
	"  -v, --verbose       Increase the verbosity.\n"
	"  -h, --help          Show this help message.\n"
	"  --autoplay          Perform a perfect run.\n"
	"  --pause             Start the game paused.\n"
	"\n"
	"Check the man page oshu(1) for details.\n"
;

static struct oshu_game current_game;

static void signal_handler(int signum)
{
	oshu_stop_game(&current_game);
}

int run(const char *beatmap_path, int autoplay, int pause)
{
	int rc = 0;

	if (SDL_Init(SDL_INIT_VIDEO|SDL_INIT_AUDIO) < 0) {
		oshu_log_error("SDL initialization error: %s", SDL_GetError());
		return -1;
	}

	if ((rc = oshu_create_game(beatmap_path, &current_game)) < 0) {
		oshu_log_error("game initialization failed");
		return -1;
	}

	current_game.autoplay = autoplay;
	if (pause)
		oshu_pause_game(&current_game);

	if ((rc = oshu_run_game(&current_game)) < 0)
		oshu_log_error("error while running the game, aborting");

	oshu_destroy_game(&current_game);
	SDL_Quit();
	return rc;
}

int main(int argc, char **argv)
{
	int autoplay = 0;
	int pause = 0;
	int verbosity = SDL_LOG_PRIORITY_INFO;

	for (;;) {
		int c = getopt_long(argc, argv, flags, options, NULL);
		if (c == -1)
			break;
		switch (c) {
		case OPT_VERBOSE:
			if (verbosity > SDL_LOG_PRIORITY_VERBOSE)
				verbosity--;
			break;
		case OPT_HELP:
			puts("oshu! version " VERSION);
			puts(usage);
			fputs(help, stdout);
			return 0;
		case OPT_AUTOPLAY:
			autoplay = 1;
			break;
		case OPT_PAUSE:
			pause = 1;
			break;
		default:
			fputs(usage, stderr);
			return 2;
		}
	}

	if (argc - optind != 1) {
		fputs(usage, stderr);
		return 2;
	}

	SDL_LogSetAllPriority(SDL_LOG_PRIORITY_WARN);
	SDL_LogSetPriority(SDL_LOG_CATEGORY_APPLICATION, (SDL_LogPriority) verbosity);
	av_log_set_level(verbosity <= SDL_LOG_PRIORITY_DEBUG ? AV_LOG_INFO : AV_LOG_ERROR);

	char *beatmap_path = realpath(argv[optind], NULL);
	if (beatmap_path == NULL) {
		oshu_log_error("cannot locate %s", argv[optind]);
		return 3;
	}

	char *beatmap_file = beatmap_path;
	char *slash = strrchr(beatmap_path, '/');
	if (slash) {
		*slash = '\0';
		oshu_log_debug("changing the current directory to %s", beatmap_path);
		if (chdir(beatmap_path) < 0) {
			oshu_log_error("error while changing directory: %s", strerror(errno));
			return 3;
		}
		beatmap_file = slash + 1;
	}

	signal(SIGTERM, signal_handler);
	signal(SIGINT, signal_handler);

	if (run(beatmap_file, autoplay, pause) < 0)
		return 1;

	free(beatmap_path);

	return 0;
}
